cJSON.c

```diff
@@ -2988,3 +2988,148 @@ CJSON_PUBLIC(void) cJSON_free(void *object)
+        if (errorMessage != NULL)
+        {
+            sprintf(errorMessage, "Cannot open the file '%.420s' to write.", filename);
+        }
+        return -1;
+    }
+
+    /* Print out the json string to the file */
+    json_str = cJSON_Print(item);
+    if (json_str != NULL)
+    {
+        fprintf(fptr, "%s", json_str);
+        fprintf(fptr, "\n");
+        free(json_str);
+    }
+    else
+    {
+        if (errorMessage != NULL)
+        {
+            sprintf(errorMessage, "Failed to print json object to a string.");
+        }
+        fclose(fptr);
+        return -2;
+    }
+    fclose(fptr);
+
+    /* Success */
+    return 1;
+}
+
+/* This will load and parse a json file.
+
+ Args:
+   filename - the name of the file to load
+   **item - a ptr to the cJSON structure to store the parsed root item in
+   **errorMessage - an optional field to return an error message
+
+ Returns:
+   1 - on success
+   0 - file not found
+  -1 - on intial memory allocation failure
+  -2 - on failure to reallocate more memory
+  -3 - bad filename arg
+  -4 - bad item arg
+*/
+#define cJSON_LOAD_BUFFER_INCREMENT 2048
+CJSON_PUBLIC(int) cJSON_loadJSONfile(const char *filename, cJSON **item, char *errorMessage) {
+
+    FILE *fptr;
+    char *buffer;
+    size_t bufferSize = cJSON_LOAD_BUFFER_INCREMENT;
+    size_t bytesRead = 0;
+    int i = 0;
+
+    /* Sanity check the item arg*/
+    if (item == NULL) {
+        return -4;
+    }
+
+    /* Sanity check the filename arg */
+    if (filename == NULL) {
+        if (errorMessage != NULL)
+        {
+            sprintf(errorMessage, "The filename was NULL");
+        }
+        return -3;
+    }
+
+    /* Open the file */
+    fptr = fopen(filename, "r");
+    if (fptr == NULL) {
+        if (errorMessage != NULL)
+        {
+            sprintf(errorMessage, "Cannot open the file '%s.420' to read.", filename);
+        }
+        return 0;
+    }
+
+    /* Allocate the buffer to it's initial size */
+    buffer = (char *)malloc(bufferSize + cJSON_LOAD_BUFFER_INCREMENT);
+    if (buffer == NULL) {
+        if (errorMessage != NULL)
+        {
+            sprintf(errorMessage, "Memory allocation error while reading '%s.420'.", filename);
+        }
+        return -1;
+    }
+    /* Read in the file until there's nothing left to read
+    Keep reading in the file, expanding our buffer if needed */
+    while ((bytesRead = fread(buffer + (cJSON_LOAD_BUFFER_INCREMENT * i), 1, cJSON_LOAD_BUFFER_INCREMENT, fptr)) > 0) {
+        if (bytesRead == sizeof(buffer))
+        {
+            /* We filled the buffer, realloc the final buffer so it's larger */
+            buffer = (char *)realloc(buffer, bufferSize + cJSON_LOAD_BUFFER_INCREMENT);
+            if (buffer == NULL)
+            {
+                if (errorMessage != NULL)
+                {
+                    sprintf(errorMessage, "Memory allocation error while reading '%.420s'.", filename);
+                }
+                return -2;
```

Comment:
Forget close file and memory should be free

Reply:
I've added the missing fclose, thank you.

'buffer' is already NULL in this case (memory allocation error), so we can't free it on this exit.

Thanks!

Reply:
If realloc() fails the original block is left untouched; it is not freed or moved. See [https://linux.die.net/man/3/realloc](url)

Reply:
Yes, you are entirely correct. I've updated to ensure the original block is freed. Thanks!